Implement minor and major releases

mathias-baumann-sociomantic opened the pull request.

leandro-lucarella-sociomantic commented on src/release/main.d:
```diff
@@ -110,6 +117,35 @@ void main ( string[] params )
+    import std.stdio;
+
+    writefln(msg);
+    writefln("Exit the shell when you are done.");
```
Maybe mention this can be done via `Ctrl-D` or by typing `exit` usually?

leandro-lucarella-sociomantic commented on src/release/options.d:
```diff
@@ -52,8 +55,11 @@ Options parseOpts ( string[] opts )
-           "verbose", "Set logging to verbose", &verbose);
+           "verbose", "Set logging to verbose", &verbose,
 
+           "release_subsequent|f",
```
Sorry for the ignorance, but will this translate to `--release-subsequent` or will it be `--release_subsequent`? If so please please make it `--release-subsequent` :)

Also, I guess that `f` is a short option (`-f`), why `f`? That usually means `--force` and there is no `f` at all in the whole `release_subsequent`...

Also, maybe `--release-all` / `-a`?

mathias-baumann-sociomantic replied:
I chose `-f` short for `follow`, as in, follow the merges and release. I'll go with release all

leandro-lucarella-sociomantic commented on src/release/main.d:
```diff
@@ -142,6 +141,7 @@ void letUserResolveConflicts ( string msg )
     if (shell_cmd.length == 0)
         shell_cmd = "bash";
 
+
```
Is this extra space intended? I don't mind, just curious...

mathias-baumann-sociomantic replied:
Yeah, it separates the preparation of the variable from the usage

leandro-lucarella-sociomantic commented on src/release/options.d:
```diff
@@ -0,0 +1,69 @@
+    auto help_info = getopt(opts,
+           "log|l", format("Set the logging level, one of %s",
+                           iota(LogLevel.min, LogLevel.max).map!LogLevel), &options.logging,
+           "verbose", "Set logging to verbose", &verbose);
```
I guess `-v` for verbose would be also nice.

mihails-strasuns-sociomantic commented on src/release/main.d:
```diff
@@ -227,6 +228,7 @@ ActionList preparePatchRelease ( ref HTTPConnection con, ref Repository repo,
         repo = repo object
         tags = array of tags for this repo
         minor_version = version to release
+        follow = is set, will also release merged branches
```
What is the intended use case for setting it to "no"?

mathias-baumann-sociomantic replied:
For when one wants to release only one.. release

mihails-strasuns-sociomantic commented on src/release/main.d:
```diff
@@ -335,6 +332,28 @@ ActionList prepareMinorRelease ( ref HTTPConnection con, ref Repository repo,
+
+*******************************************************************************/
+
+ActionList clearReleaseNotes ( )
```
I wonder if it would be better to clean release notes right _before_ tagging release. Would make merging of release notes more correct. CC @gavin-norman-sociomantic

gavin-norman-sociomantic replied:
I suppose it depends how we want the release notes to look. If we want the notes for subsequent majors to just say "Also includes merge of vx.y.0 into va.b.x", then that would be the way to do it.

That approach sounds good to me.

mihails-strasuns-sociomantic replied:
Yeah, that is what we have been doing so far (though in wording of "base on vX.Y.Z release of previous major")

mathias-baumann-sociomantic replied:
I don't think it's a good approach.. that way everyone who wants to know have to follow tons of redirections to find out what exactly has been changed

gavin-norman-sociomantic replied:
Yeah, this is the downside. I can see pros to both approaches.

mihails-strasuns-sociomantic replied:
You mean if you want to check it via submodule? It is still not that hard, pretty much `cd submodules/stuff; git show HEAD`. I wonder though what approach is more common - that or reading notes in web release page/mail.

Problem with keeping notes is that we have to either manually separate them after merging into next branch or merge everything twice - first actual released tag and "clean notes" branch commit right after.

gavin-norman-sociomantic replied:
Right. This is the hassle.

mihails-strasuns-sociomantic commented on src/release/main.d:
```diff
@@ -354,6 +354,64 @@ ActionList clearReleaseNotes ( )
+                                      format("Checkout original branch %s",
+                                             current_branch));
+    return list;
+}
```
This whole function doesn't make a lot of sense tbh. I suggested to not support major release by `neptune-release` exactly because it is mostly manual effort with very few automation. For example, this function does not rebase major branch before releasing.

Maybe it should just blindly tag / aggregate release notes on current branch trusting developer to do all necessary preparations?

mihails-strasuns-sociomantic commented on src/release/main.d:
```diff
@@ -102,6 +102,10 @@ void main ( string[] params )
     }
 
+    // Make sure we mention each ref only once or git push will complain
+    list.affected_refs.sort;
```
Please don't use parens-less calls for stateful methods at least :/

mathias-baumann-sociomantic commented:
Addressed all comments with my :+1:. Did not address the major release issue and the release notes issue because I want to talk in more detail about that in the coming meeting

mathias-baumann-sociomantic commented:
I pushed the sanity checks we discussed (extra commit)
I also added packing.

nemanja-boric-sociomantic commented on src/release/main.d:
```diff
@@ -383,6 +387,37 @@ ActionList prepareMajorRelease ( ref HTTPConnection con, ref Repository repo,
+        if (merges.length > 0)
+        {
+            import std.stdio;
+            writefln("Found merges between this and the last major "~
```
There's no way to pass a flag to allow this behavior? Not sure if needed, but it sounds harsh.

mihails-strasuns-sociomantic replied:
Better to add it when it becomes necessary than try to make everything configurable in my opinion.
